ttnn/cpp/ttnn/tensor/tensor_utils.cpp

```diff
@@ -213,6 +213,234 @@ Tensor convert_conv_weight_tensor_to_tiled_layout(
+        auto output_buffer = owned_buffer::create<T>(compute_volume(output_shape));
+        for (auto ic = 0; ic < num_channel_shards; ic++) {
+            for (auto r = 0; r < w_shape[2]; r++) {
+                for (auto s = 0; s < w_shape[3]; s++) {
+                    for (auto c_s = 0; c_s < conv_input_shard_width; c_s++) {
+                        for (auto oc = 0; oc < num_channel_shards; oc++) {
+                            for (auto k_s = 0; k_s < conv_output_shard_width; k_s++) {
+                                auto matrix_idx = (oc * conv_output_shard_width_padded + k_s) +
+                                                  c_s * weight_matrix_cols +
+                                                  s * conv_input_shard_width * weight_matrix_cols +
+                                                  r * w_shape[3] * conv_input_shard_width * weight_matrix_cols +
+                                                  ic * weight_block_height_padded * weight_matrix_cols;
+                                auto idx = (oc * conv_output_shard_width + k_s) * w_shape[1] * w_shape[2] * w_shape[3] +
+                                           (ic * conv_input_shard_width + c_s) * w_shape[2] * w_shape[3] +
+                                           r * w_shape[3] + s;
+                                output_buffer[matrix_idx] = input_buffer[idx];
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        if constexpr (std::is_same<T, float>::value) {
+            if (output_dtype == DataType::BFLOAT8_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp8_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
+            }
+            if (output_dtype == DataType::BFLOAT4_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp4_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
+            }
```

Comment:
This looks like the same code copied/pasted again and again in this file. Can you please refactor this?

Reply:
Done.